### ISSUE TITLE
[Suggesting] Crafttweaker integration for 1.16

### ISSUE THREAD

**Issue by Algester** [HellFirePvP/AstralSorcery#1715]
uhhh I think this should be possible

**Comment by PocketfulOfGeese** [comment]
What isn't already possible? Everything in 1.16 ties into the vanilla system for datapack accessibility, so unlike 1.12 everything should be accessible to craft tweaker already.

**Comment by Algester** [comment]
The serializer for the astral altar is different for example
the recipe says its type: astralsorcery:altar, but the recipetype is astralsorcery:simple_altar, but if you looked at the github recipes they arent one and the same thus fudging the logic for craft tweaker

Right now I'm just studying how to make the stuff work with craft tweaker (its simpler that way) now makes me wonder if I can "rip" the recipetype format with craft tweaker and see how I can possibly change or add recipes

**Comment by PocketfulOfGeese** [comment]
Yes, it would be the same way you modify vanilla recipes...because everything in AS ties into that system now for datapack support. Additional support to accomplish the same task isn't really a good use of dev cycles and won't be done unless something changes under the hood for AS.

**Comment by Algester** [comment]
I'll just toy around I think I can make it work with crafttweaker its just complaining how "recipe:astralsorcery:altar is not a thing and is rejecting recipe:astralsorcery:simple_altar as well

**Comment by Algester** [comment]
to give a perspective is that I'm adding a JSON recipe (which is ideally a datapack thing)

<recipetype:astralsorcery:simple_altar>.addJSONRecipe("stone2marbletest", {
  "type": "astralsorcery:altar",
  "altar_type": 0,
  "duration": 5,
  "starlight": 1,
  "pattern": [
    "_____",
    "_____",
    "__A__",
    "_____",
    "_____"
  ],
  "key": {
    "A": {
      "tag": "minecraft:stone"
    },
  },
  "output": [
    {
      "item": "astralsorcery:marble_raw",
      "count": 1
    }
  ],
  "effects": [
    "astralsorcery:built_in_effect_discovery_central_beam",
  ]
});

just to test the logic out and its throwing me cannot override recipe because it expects recipe:astralsorcery:altar to be recipe:astralsorcery:simple_altar, but if I change it to astralsorcery:simple_altar crafttweaker is throwing me recipe is invalid because its not a valid recipetype

**Comment by BlueAgent** [comment]
The recipe type id is `astralsorcery:simple_altar` and the recipe serializer id is `astralsorcery:altar` so you actually have the correct script.

CraftTweaker puts a restriction that requires the[ **recipe type** id and the and **recipe serializer** id to be the same](https://github.com/CraftTweaker/CraftTweaker/blob/e11d528ee6a1e5fce57711d28c024d7e6fc27e49/src/main/java/com/blamejared/crafttweaker/api/managers/IRecipeManager.java#L67-L70).
This is quite strange because in vanilla recipes, crafting table recipes use the recipe type `minecraft:crafting` and has many recipe serializers, two of which are `minecraft:crafting_shaped` and `minecraft:crafting_shapeless`.

~~I think this is probably okay to report this to CraftTweaker instead and maybe link back here.~~
I have made a PR to fix this here https://github.com/CraftTweaker/CraftTweaker/pull/1166.

Doesn't seem to need any changes on Astral Sorcery's end.

<details>
<summary>For reference purposes, a formatted version of your script</summary>

```zs
<recipetype:astralsorcery:simple_altar>.addJSONRecipe("stone2marbletest", {
    "type": "astralsorcery:altar",
    "altar_type": 0,
    "duration": 5,
    "starlight": 1,
    "pattern": [
        "_____",
        "_____",
        "__A__",
        "_____",
        "_____"
    ],
    "key": {
        "A": {
            "tag": "forge:stone"
        }
    },
    "output": [{
        "item": "astralsorcery:marble_raw",
        "count": 1
    }],
    "effects": [
        "astralsorcery:built_in_effect_discovery_central_beam"
    ]
});
```
</details>

<details>
<summary>Vanilla Crafting Table recipe that should work but fails</summary>

```zs
<recipetype:minecraft:crafting>.addJSONRecipe("apple_from_stone", {
    "type": "minecraft:crafting_shapeless",
    "ingredients": [
        {
            "tag": "minecraft:stone_tool_materials"
        }
    ],
    "result": {
        "item": "minecraft:apple",
        "count": 1
    }
});
```
</details>

I tested with KubeJS and it successfully adds the recipe: 
<details>
<summary>kubejs/server_scripts/test.js</summary>

```js
onEvent('recipes', event => {
    event.custom({
        "type": "astralsorcery:altar",
        "altar_type": 0,
        "duration": 5,
        "starlight": 1,
        "pattern": [
            "_____",
            "_____",
            "__A__",
            "_____",
            "_____"
        ],
        "key": {
            "A": {
                "tag": "forge:stone"
            }
        },
        "output": [{
            "item": "astralsorcery:marble_raw",
            "count": 1
        }],
        "effects": [
            "astralsorcery:built_in_effect_discovery_central_beam"
        ]
    });
});
```
</details>

**Comment by Algester** [comment]
According to Jared he's already contacted hellfire so..... I'll just wait how it works out

**Comment by BlueAgent** [comment]
Yeah I heard about that and it'll be nice.
Till then, the [PR](https://github.com/CraftTweaker/CraftTweaker/pull/1166) just got merged and is available in [CraftTweaker v7.1.0.124](https://www.curseforge.com/minecraft/mc-mods/crafttweaker/files/3204396
).
I tested adding an altar recipe with the following script and it works now 😊:

<details>
<summary>Script and Screenshots</summary>

```zs
<recipetype:astralsorcery:simple_altar>.addJSONRecipe("stone2marbletest", {
    "type": "astralsorcery:altar",
    "altar_type": 0,
    "duration": 5,
    "starlight": 1,
    "pattern": [
        "_____",
        "_____",
        "__A__",
        "_____",
        "_____"
    ],
    "key": {
        "A": {
            "tag": "forge:stone"
        }
    },
    "output": [{
        "item": "astralsorcery:marble_raw",
        "count": 1
    }],
    "effects": [
        "astralsorcery:built_in_effect_discovery_central_beam"
    ]
});
```

![Luminous Crafting Table recipe for Marble from Stones](https://user-images.githubusercontent.com/630920/107899608-b0086c00-6f79-11eb-8fdf-65883b0771e2.png)
![Recipe in the table](https://user-images.githubusercontent.com/630920/107899777-3e7ced80-6f7a-11eb-8b54-c112b269a127.png)
![Recipe mid-crafting in UI](https://user-images.githubusercontent.com/630920/107899784-4341a180-6f7a-11eb-831d-9b9681bdb7dd.png)
![Recipe mid-crafting in world](https://user-images.githubusercontent.com/630920/107899791-46d52880-6f7a-11eb-949b-788040adfb8d.png)
</details>

This issue can probably be either be closed, or left open till Jared's PR for a nicer CT api for Astral.
